### PR TITLE
Add overloads for C# Interface to use flat vector with `unsafe` to speedup passing rays

### PR DESCRIPTION
**Pull request by mariuszhermansdorfer**
This PR has a bunch of minor optimizations to the raycasting API:

- Add an overload for the Embree raytracer to accept float[] as input
- Use unsafe methods to iterate over arrays for faster conversions

Test results:
image

### CODE REVIEW

File: src/Csharp/CMakeLists.txt [hunk 1]
```diff
@@ -69,7 +69,7 @@ set_target_properties(
 		DHARTAPICSharp
 		PROPERTIES
 		VS_DOTNET_REFERENCES "${normal_references}"
-		VS_DOTNET_TARGET_FRAMEWORK_VERSION "v4.5"
+		VS_DOTNET_TARGET_FRAMEWORK_VERSION "v4.8"
```
Comment (cadop) [inline src/Csharp/CMakeLists.txt:72]: I updated the build instructions to mention this .net version issue https://github.com/cadop/dhart/pull/43

I would prefer to wait a little on incrementing this framework version until I can test the built dll's in different contexts (e.g. Unity).  

Could you revert the cmake changes for this PR?

File: src/Csharp/packages/base/src/NativeUtils.cs [hunk 2]
```diff
@@ -180,7 +181,25 @@ public static float[] FlattenVectorArray(IEnumerable<Vector3D> vectors)
             The number of elements in result_array is not equal to the number of
             elements in float_array / 3
        */
-		public static Vector3D[] FloatArrayToVectorArray(float[] float_array, bool[] result_array)
+
+        public static float[] FlattenVectorArrayUnsafe(Vector3D[] vectors)
```
Comment (cadop) [inline src/Csharp/packages/base/src/NativeUtils.cs:185]: Could you confirm that the unittests are passing after you made this change? I haven't looked at this much yet, but it seems to be changing the API?

**Review by cadop — CHANGES_REQUESTED** [review]

File: src/Csharp/packages/raytracer/src/RayTracerNative.cs [hunk 3]
```diff
@@ -16,7 +16,7 @@ namespace DHARTAPI.RayTracing
 {
 
 	/*! \param NativeMethods for the RayTracing namespace */
-	internal static class NativeMethods
+	internal sealed class NativeMethods
```
Comment (cadop) [inline src/Csharp/packages/raytracer/src/RayTracerNative.cs:19]: Why is this changed to `sealed`?

Comment (mariuszhermansdorfer) [reply src/Csharp/packages/raytracer/src/RayTracerNative.cs:19]: Performance reasons. Unless you want to derive from a class, it's generally a good idea to seal classes in c#
https://www.youtube.com/watch?v=d76WWAD99Yo

File: src/Csharp/packages/base/src/NativeUtils.cs (hunk 2 above)
Comment (mariuszhermansdorfer) [reply src/Csharp/packages/base/src/NativeUtils.cs:185]: Ah, my bad. I wanted to paste this under this method:
https://github.com/cadop/dhart/blob/44d2d9a52a18c4355c226116522bd7be99150f1a/src/Csharp/packages/base/src/NativeUtils.cs#L137-L160

I'm keeping it untouched and just introducing another overload working with Vector3D[]
image

I'll put it in the right place

File: src/Csharp/CMakeLists.txt (hunk 1 above)
Comment (mariuszhermansdorfer) [reply src/Csharp/CMakeLists.txt]: Changed in the latest commit

File: src/Csharp/packages/raytracer/src/RayTracerNative.cs (hunk 3 above)
Comment (cadop) [reply src/Csharp/packages/raytracer/src/RayTracerNative.cs:19]: I get the point he is making against regular classes, but we are using `static`. 

https://social.technet.microsoft.com/wiki/contents/articles/21028.difference-between-static-class-sealed-class-and-abstract-class-in-c.aspx

**Static Class:** ... This class **cannot be instantiated,** i.e we cannot have objects of this class. To access methods of this class, you can directly use classname.method. Also this class cannot be inherited.

**Sealed Class**:.... this class cannot be inherited. But **we can instantiate this class**, i.e we can have any number of objects of a sealed class.

So from the API standpoint, is there a reason you need to instantiate the class? I think this might be dangerous in the way we are interfacing with C, but i'm not sure. 


https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/static-classes-and-static-class-members

> Static classes are sealed and therefore cannot be inherited.

Comment (mariuszhermansdorfer) [reply src/Csharp/packages/raytracer/src/RayTracerNative.cs:19]: Good point!
I’ll revert these back to static.

**Comment by cadop** [comment]
Is there something in your gitignore or a missing commit? I can't build this branch

image

**Comment by mariuszhermansdorfer** [comment]
Try building it now @cadop.

I reverted all classes back to static and added the missing reference to CMakeLists. Previously I had it added manually here:
image

**Comment by cadop** [comment]
When you get a chance could you edit the original comment for this PR and paste a screenshot of the `hfcsharptest` group all passing.  I need to setup the automated testing on github runners, but until that happens, i'll just need to have people post that the unittests pass, and then ill review the code and test on other platforms.

**Comment by mariuszhermansdorfer** [comment]
> When you get a chance could you edit the original comment for this PR and paste a screenshot of the `hfcsharptest` group all passing. I need to setup the automated testing on github runners, but until that happens, i'll just need to have people post that the unittests pass, and then ill review the code and test on other platforms.

Updated the initial comment and added a screenshot of all tests passing.

**Comment by cadop** [comment]
Didn't mean to close this, it was because I deleted the other branch when merging it. You can rebase on main and reopen this PR to be merged into main (and ill change it).  

* You shouldn't need to make a new PR, just reopen this one
